File: src/drivers/resolve.ts
import { existsSync } from "node:fs";
import { resolve } from "node:path";
import type { Driver } from "./types";
import type { DriverConnection } from "./types";
import type { Connection } from "../lib/config";
import { getConnection, getConnections, getDefaultConnectionAlias } from "../lib/config";
import { getCredential, type Credential } from "../lib/credentials";
import { setActiveDriver, clearActiveDriver } from "../lib/cleanup";
import { connectPg } from "./pg";
import { connectSqlite } from "./sqlite";
import { connectMysql } from "./mysql";
import { connectSnowflake } from "./snowflake";

type ResolveOpts = {
  connection?: string;
  write?: boolean;
};

const DRIVER_URL_PATTERNS: [RegExp, Driver][] = [
  [/^postgres(ql)?:\/\//, "pg"],
  [/^mysql:\/\//, "mysql"],
  [/^mariadb:\/\//, "mysql"],
  [/^sqlite:\/\//, "sqlite"],
  [/^snowflake:\/\//, "snowflake"],
];

export const SQLITE_FILE_EXTENSIONS = [".sqlite", ".db", ".sqlite3", ".db3"];

export const detectDriverFromUrl = (url: string): Driver | undefined => {
  for (const [pattern, driver] of DRIVER_URL_PATTERNS) {
    if (pattern.test(url)) {
      return driver;
    }
  }

  const lower = url.toLowerCase();
  if (SQLITE_FILE_EXTENSIONS.some((ext) => lower.endsWith(ext))) {
    return "sqlite";
  }

  return undefined;
};

const resolveAlias = (explicit?: string): string => {
  if (explicit) {
    return explicit;
  }

  const envAlias = process.env.AGENT_SQL_CONNECTION?.trim();
  if (envAlias) {
    return envAlias;
  }

  const defaultAlias = getDefaultConnectionAlias();
  if (defaultAlias) {
    return defaultAlias;
  }

  const available = Object.keys(getConnections());
  const listing = available.length > 0 ? available.join(", ") : "(none configured)";
  throw Object.assign(
    new Error(
      `No connection specified and no default configured. Available connections: ${listing}. ` +
        "Tip: -c also accepts file paths (e.g. ./data.db) and connection URLs (e.g. postgres://user:pass@host/db).",
    ),
    { fixableBy: "agent" as const },
  );
};

const resolveDriverType = (conn: Connection): Driver => {
  if (conn.driver) {
    return conn.driver;
  }

  if (conn.url) {
    const detected = detectDriverFromUrl(conn.url);
    if (detected) {
      return detected;
    }
  }

  throw new Error(
    "Cannot determine driver type. Set the 'driver' field on the connection or use a URL with a recognizable scheme (postgres://, sqlite://, mysql://).",
  );
};

const checkWritePermission = (opts: {
  driver: Driver;
  credential: Credential | null;
  write: boolean;
  alias: string;
}): void => {
  if (!opts.write) {
    return;
  }

  if (opts.credential && !opts.credential.writePermission) {
    throw Object.assign(
      new Error(
        `Write mode requested but credential for connection '${opts.alias}' has writePermission disabled.`,
      ),
      {
        hint: "Update the credential with --write to enable write permission, or use a different credential.",
        fixableBy: "human" as const,
      },
    );
  }

  if (
    (opts.driver === "pg" || opts.driver === "mysql" || opts.driver === "snowflake") &&
    !opts.credential
  ) {
    const driverName =
      opts.driver === "pg" ? "PostgreSQL" : opts.driver === "mysql" ? "MySQL" : "Snowflake";
    throw Object.assign(
      new Error(
        `Write mode requested but ${driverName} connection '${opts.alias}' has no credential. ${driverName} requires a credential with writePermission to enable writes.`,
      ),
      {
        hint: "Add a credential with: agent-sql credential add <name> --username <user> --password <pass> --write",
        fixableBy: "human" as const,
      },
    );
  }
};

const trackDriver = (connection: DriverConnection): DriverConnection => {
  setActiveDriver(connection);
  const originalClose = connection.close.bind(connection);
  return {
    ...connection,
    close: async () => {
      clearActiveDriver();
      return originalClose();
    },
  };
};

export const isConnectionUrl = (value: string): boolean =>
  DRIVER_URL_PATTERNS.some(([pattern]) => pattern.test(value));

export const isFilePath = (value: string): boolean => {
  const lower = value.toLowerCase();
  if (SQLITE_FILE_EXTENSIONS.some((ext) => lower.endsWith(ext))) {
    return true;
  }
  if (value.startsWith("/") || value.startsWith("./") || value.startsWith("../")) {
    return existsSync(resolve(value));
  }
  return false;
};

const parseConnectionUrl = (
  url: string,
): { host: string; port: number; database: string; username: string; password: string } => {
  const parsed = new URL(url);
  return {
    host: parsed.hostname || "localhost",
    port: parsed.port ? Number(parsed.port) : 0,
    database: parsed.pathname.replace(/^\//, "") || "",
    username: decodeURIComponent(parsed.username),
    password: decodeURIComponent(parsed.password),
  };
};

const rejectAdHocUrlWrite = (write: boolean): void => {
  if (!write) {
    return;
  }
  throw Object.assign(
    new Error(
      "Write mode is not available for ad-hoc connections. Set up a named connection with a write-enabled credential to use --write.",
    ),
    { fixableBy: "human" as const },
  );
};

const resolveAdHocConnection = async (
  connectionStr: string,
  write: boolean,
): Promise<DriverConnection | undefined> => {
  // URL check first — more recognizable shape than file paths
  if (isConnectionUrl(connectionStr)) {
    const driver = detectDriverFromUrl(connectionStr);
    if (!driver) {
      return undefined;
    }

    if (driver === "sqlite") {
      rejectAdHocUrlWrite(write);
      const filePath = connectionStr.replace(/^sqlite:\/\//, "");
      return trackDriver(await connectSqlite({ path: resolve(filePath), readonly: true }));
    }

    if (driver === "snowflake") {
      rejectAdHocUrlWrite(write);
      const token = process.env.AGENT_SQL_SNOWFLAKE_TOKEN;
      if (!token) {
        throw Object.assign(
          new Error(
            "Ad-hoc Snowflake connections require AGENT_SQL_SNOWFLAKE_TOKEN environment variable.",
          ),
          { fixableBy: "human" as const },
        );
      }
      const parsed = new URL(connectionStr);
      const pathParts = parsed.pathname.replace(/^\//, "").split("/");
      return trackDriver(
        await connectSnowflake({
          account: parsed.hostname,
          database: pathParts[0] || undefined,
          schema: pathParts[1] || undefined,
          warehouse: parsed.searchParams.get("warehouse") ?? undefined,
          role: parsed.searchParams.get("role") ?? undefined,
          token,
          readonly: true,
        }),
      );
    }

    rejectAdHocUrlWrite(write);
    const urlParts = parseConnectionUrl(connectionStr);
    const defaultPort = driver === "pg" ? 5432 : 3306;
    const defaultDb = driver === "pg" ? "postgres" : "mysql";
    const connectOpts = {
      host: urlParts.host,
      port: urlParts.port || defaultPort,
      database: urlParts.database || defaultDb,
      username: urlParts.username,
      password: urlParts.password,
      readonly: true,
    };

    if (driver === "pg") {
      return trackDriver(await connectPg(connectOpts));
    }
    return trackDriver(await connectMysql(connectOpts));
  }

  // File path check — SQLite file, write allowed via --write
  if (isFilePath(connectionStr)) {
    const absPath = resolve(connectionStr);
    const fileExists = await Bun.file(absPath).exists();
    if (!fileExists && !write) {
      throw Object.assign(new Error(`SQLite database not found: ${connectionStr}`), {
        hint: "Check the file path, or use --write to create a new database.",
        fixableBy: "agent" as const,
      });
    }
    return trackDriver(
      await connectSqlite({ path: absPath, readonly: !write, create: !fileExists }),
    );
  }

  return undefined;
};

type ConfigConnectOpts = {
  conn: Connection;
  credential: Credential | null;
  readonly: boolean;
  alias: string;
};

const connectPgFromConfig = async (opts: ConfigConnectOpts): Promise<DriverConnection> => {
  if (!opts.credential?.username || !opts.credential?.password) {
    throw Object.assign(
      new Error(
        `PostgreSQL connection '${opts.alias}' requires a credential with username and password.`,
      ),
      {
        hint: "Add a credential with: agent-sql credential add <name> --username <user> --password <pass>",
        fixableBy: "human" as const,
      },
    );
  }

  return connectPg({
    host: opts.conn.host ?? "localhost",
    port: opts.conn.port ?? 5432,
    database: opts.conn.database ?? "postgres",
    username: opts.credential.username,
    password: opts.credential.password,
    readonly: opts.readonly,
  });
};

const connectSqliteFromConfig = async (opts: ConfigConnectOpts): Promise<DriverConnection> => {
  const path = opts.conn.path ?? opts.conn.url?.replace(/^sqlite:\/\//, "");
  if (!path) {
    throw new Error(
      `SQLite connection '${opts.alias}' requires a path. Set 'path' on the connection or use a sqlite:// URL.`,
    );
  }

  return connectSqlite({ path, readonly: opts.readonly });
};

const connectMysqlFromConfig = async (opts: ConfigConnectOpts): Promise<DriverConnection> => {
  if (!opts.credential?.username || !opts.credential?.password) {
    throw Object.assign(
      new Error(
        `MySQL connection '${opts.alias}' requires a credential with username and password.`,
      ),
      {
        hint: "Add a credential with: agent-sql credential add <name> --username <user> --password <pass>",
        fixableBy: "human" as const,
      },
    );
  }

  return connectMysql({
    host: opts.conn.host ?? "localhost",
    port: opts.conn.port ?? 3306,
    database: opts.conn.database ?? "mysql",
    username: opts.credential.username,
    password: opts.credential.password,
    readonly: opts.readonly,
  });
};

const connectSnowflakeFromConfig = async (opts: ConfigConnectOpts): Promise<DriverConnection> => {
  if (!opts.credential?.password) {
    throw Object.assign(
      new Error(
        `Snowflake connection '${opts.alias}' requires a credential with a PAT (personal access token) as the password.`,
      ),
      {
        hint: "Add a credential with: agent-sql credential add <name> --password <pat_secret>",
        fixableBy: "human" as const,
      },
    );
  }

  return connectSnowflake({
    account: opts.conn.account ?? "",
    database: opts.conn.database,
    schema: opts.conn.schema,
    warehouse: opts.conn.warehouse,
    role: opts.conn.role,
    token: opts.credential.password,
    readonly: opts.readonly,
  });
};

const configConnectBuilders: Record<
  Driver,
  (opts: ConfigConnectOpts) => Promise<DriverConnection>
> = {
  pg: connectPgFromConfig,
  sqlite: connectSqliteFromConfig,
  mysql: connectMysqlFromConfig,
  snowflake: connectSnowflakeFromConfig,
};

export const resolveDriver = async (opts?: ResolveOpts): Promise<DriverConnection> => {
  const alias = resolveAlias(opts?.connection);
  const write = opts?.write ?? false;

  // Try ad-hoc connection (URL or file path) before config lookup
  const adHoc = await resolveAdHocConnection(alias, write);
  if (adHoc) {
    return adHoc;
  }

  const conn = getConnection(alias);

  if (!conn) {
    const available = Object.keys(getConnections());
    const listing = available.length > 0 ? available.join(", ") : "(none configured)";
    throw Object.assign(
      new Error(
        `Unknown connection '${alias}'. Available connections: ${listing}. ` +
          "Tip: -c also accepts file paths (e.g. ./data.db) and connection URLs (e.g. postgres://user:pass@host/db).",
      ),
      { fixableBy: "agent" as const },
    );
  }

  const driver = resolveDriverType(conn);
  const credential = conn.credential ? getCredential(conn.credential) : null;

  // checkWritePermission throws if write is requested but not allowed
  checkWritePermission({ driver, credential, write, alias });

  const readonly = !write;

  const builder = configConnectBuilders[driver];
  if (!builder) {
    throw new Error(`Unknown driver '${driver}'. Supported drivers: pg, sqlite, mysql, snowflake.`);
  }
  return trackDriver(await builder({ conn, credential, readonly, alias }));
};

File: internal/driver/resolve.go
// Package driver contains the resolve logic for mapping connection aliases,
// URLs, and file paths to concrete driver connections.
package driver

import (
	"context"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	"github.com/shhac/agent-sql/internal/config"
	"github.com/shhac/agent-sql/internal/credential"
	"github.com/shhac/agent-sql/internal/errors"
)

// ResolveOpts configures driver resolution.
type ResolveOpts struct {
	Connection string
	Write      bool
	Timeout    int // milliseconds, 0 = use config default
}

// Resolve resolves a connection alias, URL, or file path to a Connection.
func Resolve(ctx context.Context, opts ResolveOpts) (Connection, error) {
	alias := resolveAlias(opts.Connection)

	// Try ad-hoc (URL or file path) first
	conn, err := resolveAdHoc(ctx, alias, opts.Write)
	if err != nil {
		return nil, err
	}
	if conn != nil {
		return conn, nil
	}

	// Look up in config
	return resolveFromConfig(ctx, alias, opts.Write)
}

func resolveAlias(explicit string) string {
	if explicit != "" {
		return explicit
	}
	if env := strings.TrimSpace(os.Getenv("AGENT_SQL_CONNECTION")); env != "" {
		return env
	}
	if def := config.GetDefaultAlias(); def != "" {
		return def
	}

	return ""
}

func configAliases() []string {
	conns := config.GetConnections()
	aliases := make([]string, 0, len(conns))
	for k := range conns {
		aliases = append(aliases, k)
	}
	return aliases
}

func resolveAdHoc(ctx context.Context, connStr string, write bool) (Connection, error) {
	if connStr == "" {
		return nil, errors.New(
			fmt.Sprintf("No connection specified and no default configured. Available connections: %s. Tip: -c also accepts file paths (e.g. ./data.db) and connection URLs (e.g. postgres://user:pass@host/db).",
				listAliases()),
			errors.FixableByAgent,
		)
	}

	// URL check
	if IsConnectionURL(connStr) {
		driver := DetectDriverFromURL(connStr)
		if driver == "" {
			return nil, nil
		}
		if write {
			return nil, errors.New(
				"Write mode is not available for ad-hoc connections. Set up a named connection with a write-enabled credential to use --write.",
				errors.FixableByHuman,
			)
		}
		return connectFromURL(ctx, driver, connStr)
	}

	// File path check
	if IsFilePath(connStr) {
		driver := DetectDriverFromURL(connStr)
		if driver == DriverDuckDB {
			if write {
				return nil, errors.New(
					"Write mode is not available for ad-hoc connections. Set up a named connection with a write-enabled credential to use --write.",
					errors.FixableByHuman,
				)
			}
			return connectDuckDbAdHoc(ctx, connStr)
		}
		// SQLite file
		return connectSqliteAdHoc(ctx, connStr, write)
	}

	return nil, nil
}

func resolveFromConfig(ctx context.Context, alias string, write bool) (Connection, error) {
	if alias == "" {
		return nil, errors.New(
			fmt.Sprintf("No connection specified and no default configured. Available connections: %s.",
				listAliases()),
			errors.FixableByAgent,
		)
	}

	conn := config.GetConnection(alias)
	if conn == nil {
		return nil, errors.NotFound(alias, configAliases())
	}

	driver := Driver(conn.Driver)
	if driver == "" && conn.URL != "" {
		driver = DetectDriverFromURL(conn.URL)
	}
	if driver == "" {
		return nil, errors.New(
			"Cannot determine driver type. Set the 'driver' field on the connection or use a URL with a recognizable scheme.",
			errors.FixableByAgent,
		)
	}

	// Write permission check
	if write {
		cred := getCredentialForConnection(conn)
		if err := checkWritePermission(driver, cred, alias, write); err != nil {
			return nil, err
		}
	}

	return connectFromConfig(ctx, driver, conn, !write)
}

func getCredentialForConnection(conn *config.Connection) *credential.Credential {
	if conn.Credential == "" {
		return nil
	}
	return credential.Get(conn.Credential)
}

func checkWritePermission(driver Driver, cred *credential.Credential, alias string, write bool) error {
	if !write {
		return nil
	}

	if cred != nil && !cred.WritePermission {
		return errors.New(
			fmt.Sprintf("Write mode requested but credential for connection '%s' has writePermission disabled.", alias),
			errors.FixableByHuman,
		).WithHint("Update the credential with --write to enable write permission.")
	}

	// Network drivers require a credential for writes
	needsCred := driver == DriverPG || driver == DriverCockroachDB ||
		driver == DriverMySQL || driver == DriverMariaDB ||
		driver == DriverSnowflake || driver == DriverMSSQL

	if needsCred && cred == nil {
		names := map[Driver]string{
			DriverPG: "PostgreSQL", DriverCockroachDB: "CockroachDB",
			DriverMySQL: "MySQL", DriverMariaDB: "MariaDB",
			DriverSnowflake: "Snowflake", DriverMSSQL: "MSSQL",
		}
		name := names[driver]
		return errors.New(
			fmt.Sprintf("Write mode requested but %s connection '%s' has no credential. %s requires a credential with writePermission to enable writes.", name, alias, name),
			errors.FixableByHuman,
		).WithHint("Add a credential with: agent-sql credential add <name> --username <user> --password <pass> --write")
	}

	return nil
}

func connectFromURL(ctx context.Context, driver Driver, connStr string) (Connection, error) {
	switch driver {
	case DriverSQLite:
		path := strings.TrimPrefix(connStr, "sqlite://")
		return connectSqliteAdHoc(ctx, path, false)

	case DriverDuckDB:
		path := strings.TrimPrefix(connStr, "duckdb://")
		return connectDuckDbAdHoc(ctx, path)

	case DriverPG, DriverCockroachDB:
		return connectPgLikeFromURL(ctx, driver, connStr)

	case DriverMySQL, DriverMariaDB:
		return connectMysqlLikeFromURL(ctx, driver, connStr)

	case DriverSnowflake:
		return connectSnowflakeFromURL(ctx, connStr)

	case DriverMSSQL:
		return connectMssqlFromURL(ctx, connStr)
	}
	return nil, errors.New(fmt.Sprintf("Unsupported driver: %s", driver), errors.FixableByAgent)
}

func connectFromConfig(ctx context.Context, driver Driver, conn *config.Connection, readonly bool) (Connection, error) {
	cred := getCredentialForConnection(conn)

	switch driver {
	case DriverSQLite:
		path := conn.Path
		if path == "" && conn.URL != "" {
			path = strings.TrimPrefix(conn.URL, "sqlite://")
		}
		if path == "" {
			return nil, errors.New(
				fmt.Sprintf("SQLite connection requires a path. Set 'path' on the connection or use a sqlite:// URL."),
				errors.FixableByAgent,
			)
		}
		return connectSqliteFile(ctx, path, readonly)

	case DriverDuckDB:
		path := conn.Path
		if path == "" && conn.URL != "" {
			path = strings.TrimPrefix(conn.URL, "duckdb://")
		}
		if path == "" {
			return nil, errors.New("DuckDB connection requires a path.", errors.FixableByAgent)
		}
		return connectDuckDbFile(ctx, path, readonly)

	case DriverPG, DriverCockroachDB:
		return connectPgLikeFromConfig(ctx, driver, conn, cred, readonly)

	case DriverMySQL, DriverMariaDB:
		return connectMysqlLikeFromConfig(ctx, driver, conn, cred, readonly)

	case DriverSnowflake:
		return connectSnowflakeFromConfig(ctx, conn, cred, readonly)

	case DriverMSSQL:
		return connectMssqlFromConfig(ctx, conn, cred, readonly)
	}

	return nil, errors.New(
		fmt.Sprintf("Unknown driver '%s'. Supported drivers: pg, cockroachdb, sqlite, duckdb, mysql, mariadb, snowflake, mssql.", driver),
		errors.FixableByAgent,
	)
}

func listAliases() string {
	aliases := configAliases()
	if len(aliases) == 0 {
		return "(none configured)"
	}
	return strings.Join(aliases, ", ")
}

// Helper to parse URL into host/port/database/user/password
func parseURL(connStr string) (host string, port string, database string, user string, password string, err error) {
	u, err := url.Parse(connStr)
	if err != nil {
		return "", "", "", "", "", err
	}
	host = u.Hostname()
	if host == "" {
		host = "localhost"
	}
	port = u.Port()
	database = strings.TrimPrefix(u.Path, "/")
	user = u.User.Username()
	password, _ = u.User.Password()
	return
}

// Stub connection functions — these will be populated by each driver package
// via init() or by the resolve layer importing them.
// For now, return "not implemented" errors. The actual implementations will
// be wired in when each driver agent completes.

func connectSqliteFile(ctx context.Context, path string, readonly bool) (Connection, error) {
	// Import cycle prevention: this calls into the sqlite package
	// This will be wired up properly when we have the full CLI
	return nil, fmt.Errorf("SQLite connection not wired up yet")
}

func connectSqliteAdHoc(ctx context.Context, path string, write bool) (Connection, error) {
	return nil, fmt.Errorf("SQLite ad-hoc not wired up yet")
}

func connectDuckDbAdHoc(ctx context.Context, path string) (Connection, error) {
	return nil, fmt.Errorf("DuckDB ad-hoc not wired up yet")
}

func connectDuckDbFile(ctx context.Context, path string, readonly bool) (Connection, error) {
	return nil, fmt.Errorf("DuckDB file not wired up yet")
}

func connectPgLikeFromURL(ctx context.Context, driver Driver, connStr string) (Connection, error) {
	return nil, fmt.Errorf("PG-like URL not wired up yet")
}

func connectPgLikeFromConfig(ctx context.Context, driver Driver, conn *config.Connection, cred *credential.Credential, readonly bool) (Connection, error) {
	return nil, fmt.Errorf("PG-like config not wired up yet")
}

func connectMysqlLikeFromURL(ctx context.Context, driver Driver, connStr string) (Connection, error) {
	return nil, fmt.Errorf("MySQL-like URL not wired up yet")
}

func connectMysqlLikeFromConfig(ctx context.Context, driver Driver, conn *config.Connection, cred *credential.Credential, readonly bool) (Connection, error) {
	return nil, fmt.Errorf("MySQL-like config not wired up yet")
}

func connectSnowflakeFromURL(ctx context.Context, connStr string) (Connection, error) {
	return nil, fmt.Errorf("Snowflake URL not wired up yet")
}

func connectSnowflakeFromConfig(ctx context.Context, conn *config.Connection, cred *credential.Credential, readonly bool) (Connection, error) {
	return nil, fmt.Errorf("Snowflake config not wired up yet")
}

func connectMssqlFromURL(ctx context.Context, connStr string) (Connection, error) {
	return nil, fmt.Errorf("MSSQL URL not wired up yet")
}

func connectMssqlFromConfig(ctx context.Context, conn *config.Connection, cred *credential.Credential, readonly bool) (Connection, error) {
	return nil, fmt.Errorf("MSSQL config not wired up yet")
}

// filepath.Abs wrapper for convenience
var absPath = filepath.Abs
